src/Components/ModuleForm/ModuleForm.js

```diff
@@ -0,0 +1,310 @@
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+
+      // Add logic for updating modules
+    } else if (userAction === "Update") {
+      const newModule = { moduleName, grade };
+      const updatedModuleList = moduleList.map((module) => {
+        if (module.moduleName === newModule.moduleName) {
+          return newModule;
+        } else {
+          return module;
+        }
+      });
+
+      if (updatedModuleList.length === moduleList.length) {
+        console.log("No module found");
+        this.setState({
+          error: "No module found",
+          moduleList: [...moduleList],
+        });
+        event.preventDefault();
+        return;
+      }
+
+      this.setState(
+        {
+          moduleList: updatedModuleList,
+          moduleName: "",
+          grade: "",
+          userAction: "Update",
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+
+      // Add logic for deleting modules
+    } else if (userAction === "Delete") {
+      const newModule = { moduleName, grade };
+      const updatedModuleList = moduleList.filter(
+        (module) => module.moduleName !== newModule.moduleName
+      );
+
+      if (updatedModuleList.length === moduleList.length) {
+        console.log("No module found");
+        this.setState({
+          error: "No module found",
+          moduleList: [...moduleList],
+        });
+        event.preventDefault();
+        return;
+      }
+
+      this.setState(
+        {
+          moduleList: [...updatedModuleList],
+          moduleName: "",
+          grade: this.state.grade,
+          userAction: "Delete",
+        },
+        () => {
+          localStorage.setItem(
+            "moduleList",
+            JSON.stringify(this.state.moduleList)
+          );
+        }
+      );
+      // Add logic for deleting modules
+      console.log("Delete Module");
+    }
+
+    event.preventDefault();
+  };
+
+  handleActionChange = (event) => {
+    this.setState({
+      userAction: event.target.value,
+    });
+    console.log("Action Change");
+    event.preventDefault();
+  };
+
+  calculateGPA = () => {
+    const { moduleList } = this.state;
+    const totalModules = moduleList.length;
+    if (totalModules === 0) {
+      return 0;
+    }
+    const totalGrades = moduleList.reduce((acc, module) => {
+      switch (module.grade) {
+        case "A":
+          return acc + 5.0;
+        case "A-":
+          return acc + 4.5;
+        case "B+":
+          return acc + 4.0;
+        case "B":
+          return acc + 3.5;
+        case "B-":
+          return acc + 3.0;
+        case "C+":
+          return acc + 2.5;
+        case "C":
+          return acc + 2.0;
+        case "D":
+          return acc + 1.5;
```

Comment:
I think we could define a mapping of grade to gpa without the switch statement. A static list outside of the component, and then access it easily. 
```
const gradesToGPA = {
    A: 5.0,
    ["A-"]: 4.5
}

and then in the reduce: return acc + gradesToGPA[module.grade]
```

Reply:
This way the code doesn't need to traverse through the whole switch statement every iteration but has a simple object accessor